### PR TITLE
Enable SecureString (via SqlCredential) for all APIs

### LINKED ISSUE #170: SqlCredential: Support Secure Credential passing for Sql operations
The SqlConnection object supports receiving a Sql Authentication password as a SecureString wrapped in the SqlCredential object, enabling a secret to be held in managed memory as an encrypted object, which is a recommended security practice as part of a defense in depth strategy.  See ASP.Net blog titled: Safer passwords with SqlCredential

However, the ElasticScale.Client API that leverages partial Sql Connection strings to interact with both the Shard Map Manager and individual Shards, prevents a SecureString from being used with the SqlConnection object, causing any application using the ElasticScale.Client API that has to support the Sql Auth surface area, to un-encrypt SecureStrings in managed memory, before passing the secret to the ElasticScale library via the ConnectionString (rather than a secure SqlCredential, containing a SecureString).

The recommendation here is to support the SqlCredential type, as a non-breaking change, in all places where an unencrypted connection string is asked for in the API.  This could be done by either adding function overloads to all relevant public APIs (which are already heavily overloaded), or adding an optional parameter to the existing public APIs.  Is there feedback on which of these two paths would be the best fit?

i.e. As an optional parameter, the change would look like this:

From:

public static ShardMapManager CreateSqlShardMapManager(
            string connectionString,
            ShardMapManagerCreateMode createMode)
        {

public static ShardMapManager CreateSqlShardMapManager(
            string connectionString,
            ShardMapManagerCreateMode createMode,
            SqlCredential secureCredential = null)
        {

### PR DESCRIPTION
**Pull request by stuartpa**
Add support for the SqlCredential type, as a non-breaking change, in all places where an unencrypted connection string is asked for in the API. 

Unit-tests have been added for all the new overloads.

Fix #170

### CODE REVIEW

File: Src/ElasticScale.Client/ShardManagement/Errors.resx [hunk 1]
```diff
@@ -213,6 +213,9 @@
   <data name="_SqlShardMapManagerCredentials_ConnectionStringPropertyRequired" xml:space="preserve">
     <value>The required property '{0}' must be set in the connection string.</value>
   </data>
+  <data name="_SqlShardMapManagerCredentials_ConnectionStringPropertyNotAllowed" xml:space="preserve">
+    <value>The required property '{0}' must not be set in the connection string.</value>
```
Comment (jaredmoo) [inline Src/ElasticScale.Client/ShardManagement/Errors.resx]: Is 'required' correct here?

File: Src/ElasticScale.Client/ShardManagement/SqlStore/SqlShardMapManagerCredentials.cs [hunk 2]
```diff
@@ -135,24 +173,51 @@ internal static void EnsureCredentials(SqlConnectionStringBuilder connectionStri
                 return;
             }
 
-            // UserID must be set when integrated authentication is disabled.
-            if (string.IsNullOrEmpty(connectionString.UserID))
+            // If secure credential not specified, verify that user/pwd are in the connection string. If secure credential
```
Comment (jaredmoo) [inline Src/ElasticScale.Client/ShardManagement/SqlStore/SqlShardMapManagerCredentials.cs]: I assume SqlCredential validates that username & password are not null. Is that right?

File: Src/ElasticScale.Client/ShardManagement/Errors.resx (hunk 1 above)
Comment (stuartpa) [reply Src/ElasticScale.Client/ShardManagement/Errors.resx]: No it isn't, cut and paste issue from line above!

File: Src/ElasticScale.Client/ShardManagement/SqlStore/SqlShardMapManagerCredentials.cs (hunk 2 above)
Comment (stuartpa) [reply Src/ElasticScale.Client/ShardManagement/SqlStore/SqlShardMapManagerCredentials.cs]: Correct, if null is provided for either username/password to SqlCredential, it throws: "Unhandled Exception: System.ArgumentNullException: Value cannot be null."

**Comment by phekmat** [comment]
@jaredmoo any progress on merging this in and getting a new nuget package out?

**Comment by jaredmoo** [comment]
Closing Stuart's PR, reopening as #174

**Comment by jaredmoo** [comment]
@phekmat I am working on building an official nuget package, until then I suggest you can build from master branch yourself using `dotnet build` and use the resulting `.dll`s.

**Comment by phekmat** [comment]
Will do. Appreciate the quick response

**Pull request closed without merging**